src/timely-util/src/builder_async.rs

```diff
@@ -212,6 +219,110 @@ impl<'handle, T: Timestamp, D: Container, P: Pull<BundleCore<T, D>>> Future
     }
 }
 
+// TODO: delete and use CapabilityTrait instead once TimelyDataflow/timely-dataflow#512 gets merged
+pub trait CapabilityTrait<T: Timestamp> {
+    fn session<'a, D, P>(
+        &'a self,
+        handle: &'a mut OutputHandleCore<'_, T, D, P>,
+    ) -> Session<'a, T, D, PushCounter<T, D, P>>
+    where
+        D: Container,
+        P: Push<BundleCore<T, D>>;
+}
+
+impl<T: Timestamp> CapabilityTrait<T> for InputCapability<T> {
+    fn session<'a, D, P>(
+        &'a self,
+        handle: &'a mut OutputHandleCore<'_, T, D, P>,
+    ) -> Session<'a, T, D, PushCounter<T, D, P>>
+    where
+        D: Container,
+        P: Push<BundleCore<T, D>>,
+    {
+        handle.session(self)
+    }
+}
+
+impl<T: Timestamp> CapabilityTrait<T> for Capability<T> {
+    fn session<'a, D, P>(
+        &'a self,
+        handle: &'a mut OutputHandleCore<'_, T, D, P>,
+    ) -> Session<'a, T, D, PushCounter<T, D, P>>
+    where
+        D: Container,
+        P: Push<BundleCore<T, D>>,
+    {
+        handle.session(self)
+    }
+}
+
+pub struct AsyncOutputHandle<T: Timestamp, D: Container, P: Push<BundleCore<T, D>> + 'static> {
+    handle: Rc<RefCell<OutputHandleCore<'static, T, D, P>>>,
+    wrapper: Rc<Pin<Box<OutputWrapper<T, D, P>>>>,
+}
+
+impl<T, D, P> AsyncOutputHandle<T, D, P>
+where
+    T: Timestamp,
+    D: Container,
+    P: Push<BundleCore<T, D>> + 'static,
+{
+    fn new(wrapper: OutputWrapper<T, D, P>) -> Self {
+        let mut wrapper = Box::pin(wrapper);
+        // SAFETY:
+        // get_unchecked_mut is safe because we are not moving the wrapper
+        //
+        // transmute is safe because:
+        // * We're erasing the lifetime but we guarantee through field order that the handle will
+        //   be dropped before the wrapper, thus manually enforcing the lifetime.
+        // * We never touch wrapper again after this point
+        let handle = unsafe {
+            let handle = wrapper.as_mut().get_unchecked_mut().activate();
+            std::mem::transmute::<OutputHandleCore<'_, T, D, P>, OutputHandleCore<'static, T, D, P>>(
+                handle,
+            )
+        };
+        Self {
+            handle: Rc::new(RefCell::new(handle)),
+            wrapper: Rc::new(wrapper),
+        }
+    }
+
+    #[allow(clippy::unused_async)]
+    pub async fn give_container<C: CapabilityTrait<T>>(&mut self, cap: &C, container: &mut D) {
+        let mut handle = self.handle.borrow_mut();
+        cap.session(&mut handle).give_container(container);
+    }
+
+    fn cease(&mut self) {
+        self.handle.borrow_mut().cease()
+    }
+}
+
+impl<'a, T, D, P> AsyncOutputHandle<T, Vec<D>, P>
+where
+    T: Timestamp,
+    D: Data,
+    P: Push<BundleCore<T, Vec<D>>> + 'static,
+{
+    #[allow(clippy::unused_async)]
+    pub async fn give<C: CapabilityTrait<T>>(&mut self, cap: &C, data: D) {
+        let mut handle = self.handle.borrow_mut();
+        cap.session(&mut handle).give(data);
+    }
```

Comment:
In principal this seems fine, but I'm wondering about performance implications. `give` is supposed to be very low cost, and including the `session` call plus async introduces a significant overhead. It might be that we can't measure it at the moment, but I hope at some point we will. Because of this, I think we should build abstractions in terms of "batch of data" instead of "single item".

Reply:
Calling `session` only adds an additional time comparison per call. If the time of the capability is the same then multiple consecutive `give` calls will indeed create a batch of the data instead of emitting them in individual degenerate containers

Reply:
Also the `async` call has the same overhead as a normal function call. In other words `give{_container}(..).await` does *not* need to yield to timely to produce the message. In fact there is no await point in the implementation in this PR. But the plan is to associate output handles with fuel so that when `fuel == 0` then the handle will yield for you and ensure that data isn't being buffered up in the operator

Reply:
I should probably give a bit more context on why the sessions can't be exposed to users. A timely session borrows the buffer that is associated with an `OutputHandle` which means that while a session is active nothing can touch the output handle.

However this is exactly what I want to do in this case. The `AsyncOutputHandle` is shared between the user code (the async logic) and the operator infrastructure. And what happens is that whenever the user code yields for any reason the operator infrastructure needs to be able to call `output.cease()` on all the output handles so that data can be passed downstream.

This means that we can never give an actual timely `Session` type to user code which would potentially keep it alive across an await point, because then we'll be unable to flush the output handle. So this is the compromise made by this implementation by starting a session for every `give`/`give_container`.

Reply:
And a final point (sorry for the multiple comments) is that if you look at the diff you'll find that all the async operators were already defensively creating temporary sessions, often in separate blocks, to ensure they are never held across await points[1][2]. So this implementation just codifies a pattern that we already had to ensure manually. 

[1] https://github.com/MaterializeInc/materialize/pull/17998/files#diff-4d9165ee597139d5a552a908445a834a77fead765eb10b0b6a85f2a5b9e3ca56L479
[2] https://github.com/MaterializeInc/materialize/pull/17998/files#diff-4d9165ee597139d5a552a908445a834a77fead765eb10b0b6a85f2a5b9e3ca56L754

Reply:
I did a quick experiment where a Timely operator either calls `output.session` outside the loop of processing its data, or for every call to `give`. It's about a 2x performance degradation, and about 3.5x vs. `give_container`. This is with additional inline annotations on the `session` code path, which are currently not in Timely.

Reply:
I see; I don't want to block this work, but it seems we might want to consider performance implications at some point in the future!

Reply:
I you have your benchmark handy could you check with this patch applied?

```
diff --git a/timely/src/dataflow/channels/pushers/buffer.rs b/timely/src/dataflow/channels/pushers/buffer.rs
index 6f92dab..841bce0 100644
--- a/timely/src/dataflow/channels/pushers/buffer.rs
+++ b/timely/src/dataflow/channels/pushers/buffer.rs
@@ -36,8 +36,13 @@ impl<T, C: Container, P: Push<BundleCore<T, C>>> BufferCore<T, C, P> where T: Eq

     /// Returns a `Session`, which accepts data to send at the associated time
     pub fn session(&mut self, time: &T) -> Session<T, C, P> {
-        if let Some(true) = self.time.as_ref().map(|x| x != time) { self.flush(); }
-        self.time = Some(time.clone());
+        match self.time {
+            Some(ref t) => if t != time {
+                self.flush();
+                self.time = Some(time.clone())
+            },
+            None => self.time = Some(time.clone()),
+        };
         Session { buffer: self }
     }
     /// Allocates a new `AutoflushSession` which flushes itself on drop.
```

Reply:
I tested it within Timely, here's the instructions:
```diff
diff --git a/timely/src/dataflow/channels/pushers/buffer.rs b/timely/src/dataflow/channels/pushers/buffer.rs
index 44079f7a..ac96a2f2 100644
--- a/timely/src/dataflow/channels/pushers/buffer.rs
+++ b/timely/src/dataflow/channels/pushers/buffer.rs
@@ -35,6 +35,7 @@ impl<T, C: Container, P: Push<BundleCore<T, C>>> BufferCore<T, C, P> where T: Eq
     }
 
     /// Returns a `Session`, which accepts data to send at the associated time
+    #[inline]
     pub fn session(&mut self, time: &T) -> Session<T, C, P> {
         if let Some(true) = self.time.as_ref().map(|x| x != time) { self.flush(); }
         self.time = Some(time.clone());
@@ -62,6 +63,7 @@ impl<T, C: Container, P: Push<BundleCore<T, C>>> BufferCore<T, C, P> where T: Eq
     }
 
     /// moves the contents of
+    #[inline]
     fn flush(&mut self) {
         if !self.buffer.is_empty() {
             let time = self.time.as_ref().unwrap().clone();
diff --git a/timely/src/dataflow/operators/capability.rs b/timely/src/dataflow/operators/capability.rs
index 5caa9bce..c06c4105 100644
--- a/timely/src/dataflow/operators/capability.rs
+++ b/timely/src/dataflow/operators/capability.rs
@@ -42,12 +42,14 @@ pub trait CapabilityTrait<T: Timestamp> {
 
 impl<'a, T: Timestamp, C: CapabilityTrait<T>> CapabilityTrait<T> for &'a C {
     fn time(&self) -> &T { (**self).time() }
+    #[inline]
     fn valid_for_output(&self, query_buffer: &Rc<RefCell<ChangeBatch<T>>>) -> bool {
         (**self).valid_for_output(query_buffer)
     }
 }
 impl<'a, T: Timestamp, C: CapabilityTrait<T>> CapabilityTrait<T> for &'a mut C {
     fn time(&self) -> &T { (**self).time() }
+    #[inline]
     fn valid_for_output(&self, query_buffer: &Rc<RefCell<ChangeBatch<T>>>) -> bool {
         (**self).valid_for_output(query_buffer)
     }
@@ -66,6 +68,7 @@ pub struct Capability<T: Timestamp> {
 
 impl<T: Timestamp> CapabilityTrait<T> for Capability<T> {
     fn time(&self) -> &T { &self.time }
+    #[inline]
     fn valid_for_output(&self, query_buffer: &Rc<RefCell<ChangeBatch<T>>>) -> bool {
         Rc::ptr_eq(&self.internal, query_buffer)
     }
diff --git a/timely/src/dataflow/operators/generic/handles.rs b/timely/src/dataflow/operators/generic/handles.rs
index eeddf529..851cf27a 100644
--- a/timely/src/dataflow/operators/generic/handles.rs
+++ b/timely/src/dataflow/operators/generic/handles.rs
@@ -230,6 +230,7 @@ impl<'a, T: Timestamp, C: Container, P: Push<BundleCore<T, C>>> OutputHandleCore
     ///            });
     /// });
     /// ```
+    #[inline]
     pub fn session<'b, CT: CapabilityTrait<T>>(&'b mut self, cap: &'b CT) -> Session<'b, T, C, PushCounter<T, C, P>> where 'a: 'b {
         assert!(cap.valid_for_output(&self.internal_buffer), "Attempted to open output session with invalid capability");
         self.push_buffer.session(cap.time())
diff --git a/timely/src/dataflow/operators/map.rs b/timely/src/dataflow/operators/map.rs
index 77eced3f..c27a982e 100644
--- a/timely/src/dataflow/operators/map.rs
+++ b/timely/src/dataflow/operators/map.rs
@@ -63,8 +63,11 @@ impl<S: Scope, D: Data> Map<S, D> for Stream<S, D> {
         self.unary(Pipeline, "MapInPlace", move |_,_| move |input, output| {
             input.for_each(|time, data| {
                 data.swap(&mut vector);
-                for datum in vector.iter_mut() { logic(datum); }
-                output.session(&time).give_vec(&mut vector);
+                let mut session = output.session(&time);
+                for mut datum in vector.drain(..) {
+                    logic(&mut datum);
+                    session.give(datum);
+                }
             })
         })
     }
```

In the last block, move the `let mut session` into the loop. This is what I'm observing, first with session outside the loop, then with the session inside the loop:
```
➜  timely-dataflow git:(master) ✗ cargo build --release --example pingpong; time target/release/examples/pingpong 1 500000000 -w1
    Finished release [optimized + debuginfo] target(s) in 0.05s
target/release/examples/pingpong 1 500000000 -w1  5.13s user 0.00s system 99% cpu 5.130 total
➜  timely-dataflow git:(master) ✗ cargo build --release --example pingpong; time target/release/examples/pingpong 1 500000000 -w1
   Compiling timely v0.12.0 (/home/moritz/dev/repos/timely-dataflow/timely)
    Finished release [optimized + debuginfo] target(s) in 5.89s
target/release/examples/pingpong 1 500000000 -w1  2.79s user 0.00s system 99% cpu 2.795 total
```

Reply:
With the patch I sent above I see 4.9 seconds when creating the session for every give and 3.6 seconds for creating the session once, so much smaller 1.36x slowdown. I'll send a timely PR with that patch

Reply:
Posted here https://github.com/TimelyDataflow/timely-dataflow/pull/513

Reply:
The check if the capability is valid for the output seems to be quite expensive. When removing it, the time drops from 5.1s to 3.3s. But it's there for a reason...

Reply:
Your patch doesn't change the result. I guess cloning a `Copy` timestamp is basically free.